src/Compilers/Core/Portable/SourceGenerator.cs

```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    public interface ISourceGenerator
+    {
+        void Execute(SourceGeneratorContext context);
+    }
+
+    public struct SourceGeneratorContext
+    {
+        public Compilation Compilation { get; }
+
+        // TODO: replace AnalyzerOptions with an differently named type that is otherwise identical.
+        // The concern being that something added to one isn't necessarily applicable to the other.
+        public AnalyzerOptions AnalyzerOptions { get; }
+
+        public CancellationToken CancellationToken { get; }
+
+        public void ReportDiagnostic(Diagnostic diagnostic) { throw new NotImplementedException(); }
```

Comment:
@agocke @chsienki Should we just be expecting that a generator also reporting diagnostics just reports as a regular diagnostic analyzer? Because we also don't have something like "SupportedDiganostics" which seems odd, then.

Reply:
Agreed, lets add that.